File: lld/week-02-03-patterns/facade/python/facade.py
"""Facade pattern — HomeTheaterFacade coordinates five subsystems behind two calls.

See ../README.md for the design writeup.
"""
from __future__ import annotations


class Amplifier:
    def __init__(self) -> None:
        self.on = False
        self.volume = 0

    def turn_on(self) -> None:
        self.on = True

    def turn_off(self) -> None:
        self.on = False
        self.volume = 0

    def set_volume(self, volume: int) -> None:
        self.volume = volume


class DvdPlayer:
    def __init__(self) -> None:
        self.on = False
        self.movie = ""

    def turn_on(self) -> None:
        self.on = True

    def turn_off(self) -> None:
        self.on = False
        self.movie = ""

    def play(self, movie: str) -> None:
        self.movie = movie

    def stop(self) -> None:
        self.movie = ""


class Projector:
    def __init__(self) -> None:
        self.on = False
        self.input = ""

    def turn_on(self) -> None:
        self.on = True

    def turn_off(self) -> None:
        self.on = False
        self.input = ""

    def set_input(self, source: str) -> None:
        self.input = source


class Screen:
    def __init__(self) -> None:
        self.down = False

    def lower(self) -> None:
        self.down = True

    def raise_(self) -> None:
        self.down = False


class Lights:
    def __init__(self) -> None:
        self.dimmed = False

    def dim(self) -> None:
        self.dimmed = True

    def brighten(self) -> None:
        self.dimmed = False


class HomeTheaterFacade:
    """Hides the coordination order of five subsystems behind two calls."""

    def __init__(self) -> None:
        self.amp = Amplifier()
        self.dvd = DvdPlayer()
        self.projector = Projector()
        self.screen = Screen()
        self.lights = Lights()

    def watch_movie(self, movie: str) -> None:
        self.lights.dim()
        self.screen.lower()
        self.projector.turn_on()
        self.projector.set_input("dvd")
        self.amp.turn_on()
        self.amp.set_volume(7)
        self.dvd.turn_on()
        self.dvd.play(movie)

    def end_movie(self) -> None:
        self.dvd.stop()
        self.dvd.turn_off()
        self.amp.turn_off()
        self.projector.turn_off()
        self.screen.raise_()
        self.lights.brighten()


def _demo() -> None:
    theater = HomeTheaterFacade()
    theater.watch_movie("Inception")
    print(f"Now playing: {theater.dvd.movie}")
    theater.end_movie()
    print(f"Movie ended, lights dimmed: {theater.lights.dimmed}")


if __name__ == "__main__":
    _demo()

File: lld/week-02-03-patterns/facade/go/facade.go
// Package facade demonstrates the Facade pattern: a HomeTheaterFacade
// coordinates five independent subsystems (amplifier, DVD player, projector,
// screen, lights) behind two simple calls, WatchMovie and EndMovie.
package facade

type Amplifier struct {
	On     bool
	Volume int
}

func (a *Amplifier) TurnOn()         { a.On = true }
func (a *Amplifier) TurnOff()        { a.On = false; a.Volume = 0 }
func (a *Amplifier) SetVolume(v int) { a.Volume = v }

type DvdPlayer struct {
	On    bool
	Movie string
}

func (d *DvdPlayer) TurnOn()           { d.On = true }
func (d *DvdPlayer) TurnOff()          { d.On = false; d.Movie = "" }
func (d *DvdPlayer) Play(movie string) { d.Movie = movie }
func (d *DvdPlayer) Stop()             { d.Movie = "" }

type Projector struct {
	On    bool
	Input string
}

func (p *Projector) TurnOn()               { p.On = true }
func (p *Projector) TurnOff()              { p.On = false; p.Input = "" }
func (p *Projector) SetInput(input string) { p.Input = input }

type Screen struct {
	Down bool
}

func (s *Screen) Lower() { s.Down = true }
func (s *Screen) Raise() { s.Down = false }

type Lights struct {
	Dimmed bool
}

func (l *Lights) Dim()      { l.Dimmed = true }
func (l *Lights) Brighten() { l.Dimmed = false }

// HomeTheaterFacade hides the coordination order of five subsystems behind
// two calls, so a client never needs to know the correct on/off sequence.
type HomeTheaterFacade struct {
	Amp       *Amplifier
	Dvd       *DvdPlayer
	Projector *Projector
	Screen    *Screen
	Lights    *Lights
}

func NewHomeTheaterFacade() *HomeTheaterFacade {
	return &HomeTheaterFacade{
		Amp:       &Amplifier{},
		Dvd:       &DvdPlayer{},
		Projector: &Projector{},
		Screen:    &Screen{},
		Lights:    &Lights{},
	}
}

func (f *HomeTheaterFacade) WatchMovie(movie string) {
	f.Lights.Dim()
	f.Screen.Lower()
	f.Projector.TurnOn()
	f.Projector.SetInput("dvd")
	f.Amp.TurnOn()
	f.Amp.SetVolume(7)
	f.Dvd.TurnOn()
	f.Dvd.Play(movie)
}

func (f *HomeTheaterFacade) EndMovie() {
	f.Dvd.Stop()
	f.Dvd.TurnOff()
	f.Amp.TurnOff()
	f.Projector.TurnOff()
	f.Screen.Raise()
	f.Lights.Brighten()
}
